grakn-graph/src/main/java/ai/grakn/graph/internal/AbstractGraknGraph.java

```diff
@@ -988,8 +984,14 @@ public boolean fixDuplicateCastings(String index, Set<ConceptId> castingVertexId
      * @return True if the roleplayers of the relations are the same.
      */
     private boolean relationsEqual(Relation mainRelation, Relation otherRelation){
-        return mainRelation.allRolePlayers().equals(otherRelation.allRolePlayers()) &&
-                mainRelation.type().equals(otherRelation.type());
+        String mainIndex = getRelationIndex((RelationImpl) mainRelation);
+        String otherIndex = getRelationIndex((RelationImpl) otherRelation);
+        return mainIndex.equals(otherIndex);
+    }
+    private String getRelationIndex(RelationImpl relation){
+        String index = relation.getIndex();
+        if(index == null) index = RelationImpl.generateNewHash(relation.type(), relation.allRolePlayers());
```

Comment:
The relation only gets it's index at the end of the transaction on commit so until you commit the index might be null. The reason it only gets its index on commit is because we use that same index for preventing duplicate relations and if you assign the index during relation construction you could get incorrect exceptions because of incomplete relations being incorrectly identified as duplicates.